Sup 38 create edit delete warehouse

alexeyoverchenko opened the pull request:
made create/edit/delete functional

alexeyoverchenko committed:
- in progress

alexeyoverchenko committed:
- in progress

alexeyoverchenko committed:
- return BaseRepository method to default name

alexeyoverchenko committed:
Merge branch 'develop' into sup-38_create_edit_delete_warehouse

alexeyoverchenko committed:
- in progress

alexeyoverchenko committed:
- in progress

alexeyoverchenko committed:
Merge branch 'develop' into sup-38_create_edit_delete_warehouse

# Conflicts:
#	src/main/java/by/itech/lab/supplier/controller/WarehouseController.java
#	src/main/java/by/itech/lab/supplier/domain/Warehouse.java
#	src/main/java/by/itech/lab/supplier/dto/WarehouseDto.java
#	src/main/java/by/itech/lab/supplier/dto/mapper/WarehouseMapper.java
#	src/main/java/by/itech/lab/supplier/repository/WarehouseRepository.java
#	src/main/java/by/itech/lab/supplier/service/WarehouseService.java
#	src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java

alexeyoverchenko committed:
- in progress

alexeyoverchenko committed:
- in progress

alexeyoverchenko committed:
- made save, delete, update functional

alexeyoverchenko committed:
- edited endpoint's path
- resolved save customer_id problem,now we can save this data

alexeyoverchenko committed:
- fixed updateWarehouse request (now it use overloaded method)
- deleted address entity validation

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/controller/WarehouseController.java:
```diff
@@ -6,17 +6,24 @@
 @RestController
 @AllArgsConstructor
-@RequestMapping(URL_WAREHOUSE)
+@RequestMapping(URL_CUSTOMER + URL_CUSTOMER_ID + URL_WAREHOUSE)
```
I think usage of constant here brings more issues than solves. It's hard to read and feels cumbersome.

alexeyoverchenko replied:
we should talk about it on the meeting

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/controller/WarehouseController.java:
```diff
@@ -30,4 +37,22 @@
+
+    @PutMapping(URL_ID_PARAMETER)
+    public WarehouseDto updateWarehouse(@PathVariable(name = "customerId") Long customerId, @PathVariable Long id, @Valid @RequestBody WarehouseDto warehouseDto) {
+        warehouseDto.setId(id);
```
Can we set Id to this dto on the client side?

alexeyoverchenko replied:
you mean, take data about "customer" from user's input?

alexeyoverchenko replied:
I suppose that no, cuz we should get id for comparison from our request

FortuneAwaker replied:
why we need id here, if we can put it into warehouseDto?

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/controller/WarehouseController.java:
```diff
@@ -30,4 +37,22 @@
+        return warehouseService.save(customerId, warehouseDto);
+    }
+
+    @PutMapping(URL_ID_PARAMETER)
```
If we use @ResponseStatus on some endpoints then let's use it across all application, so to be consistent.

alexeyoverchenko replied:
when I write this annotation, I concreting response status, another way, i expect status.OK, which is the default status, am I wrong?

svareyko replied:
I agreed, that we shouldnt place annotations which dont change anything

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/domain/Customer.java:
```diff
@@ -38,6 +40,7 @@
     private boolean active;
     private LocalDate deletedAt;
     @OneToMany(mappedBy = "customer")
+    @EqualsAndHashCode.Exclude
```
Can we put this  annotation on the class level? It's kind of metadata and it's easier to look at one place and understand what is excluded, otherwise you need to find it in the code.

alexeyoverchenko replied:
I think, we don't need this annotation at all

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/domain/Warehouse.java:
```diff
@@ -42,8 +44,9 @@
     private Address address;
     @ManyToOne
     @JoinColumn(name = "customer_id")
+    @EqualsAndHashCode.Exclude
```
Why do we exclude customer?

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/dto/WarehouseDto.java:
```diff
@@ -26,9 +26,6 @@
     private Double totalCapacity;
-    @Size(min = 5, max = 50, message = "Your address should contains at least 5 letters")
-    @NotBlank
     private Address address;
```
Do we need to use AddressDto here? It seems like this way we mix models and dto objects, dao level with controller level.

alexeyoverchenko replied:
Yes, but Address' task is not ready yet. and I can't use AddressDto

eugenmolchanov commented on src/main/java/by/itech/lab/supplier/dto/mapper/WarehouseMapper.java:
```diff
@@ -35,8 +34,15 @@ public WarehouseDto map(Warehouse entity) {
+        to.setType(from.getType());
+        to.setTotalCapacity(from.getTotalCapacity());
+        to.setAddress(from.getAddress());
+        to.setUsers(from.getUsersDto().stream().map(userMapper::map).collect(Collectors.toSet()));
```
Can UsersDto be null here?

alexeyoverchenko replied:
we discussed this, and make decision that we don't need this user at all. Thanks for bringing this to my attention

eugenmolchanov reviewed (changes requested):
Please, see comments.

svareyko commented on src/main/java/by/itech/lab/supplier/controller/WarehouseController.java:
```diff
@@ -30,4 +37,22 @@
+    }
+
+    @PutMapping(URL_ID_PARAMETER)
+    public WarehouseDto updateWarehouse(@PathVariable(name = "customerId") Long customerId, @PathVariable Long id, @Valid @RequestBody WarehouseDto warehouseDto) {
```
`@PathVariable(name = "customerId") Long customerId`
`name` not necessary if your variable names as expected

alexeyoverchenko replied:
Thanks!

svareyko commented on src/main/java/by/itech/lab/supplier/domain/Warehouse.java:
```diff
@@ -42,8 +44,9 @@
     @EqualsAndHashCode.Exclude
     private Set<User> users = new HashSet<>();
-}
+}
```
> No new line at end of file

svareyko commented on src/main/java/by/itech/lab/supplier/dto/WarehouseDto.java:
```diff
@@ -26,9 +26,6 @@
-    private CustomerDto customerDto;
     private Set<UserDto> usersDto = new HashSet<>();
-}
+}
```
> No new line at end of file

svareyko commented on src/main/java/by/itech/lab/supplier/dto/mapper/WarehouseMapper.java:
```diff
@@ -35,8 +34,15 @@ public WarehouseDto map(Warehouse entity) {
+        to.setAddress(from.getAddress());
+        to.setUsers(from.getUsersDto().stream().map(userMapper::map).collect(Collectors.toSet()));
+    }
+}
```
> No new line at end of file

alexeyoverchenko committed:
- deleted set<user> in entity
- deleted unnecessary annotations
- added empty lines

alexeyoverchenko committed:
Merge branch 'develop's_http_tests' into sup-38_create_edit_delete_warehouse

alexeyoverchenko committed:
- deleted set<user> in entity
- deleted unnecessary annotations
- added empty lines
- added "identifier" validation

alexeyoverchenko committed:
Merge branch 'develop' into sup-38_create_edit_delete_warehouse

# Conflicts:
#	src/main/java/by/itech/lab/supplier/domain/Address.java

alexeyoverchenko committed:
Merge branch 'develop' into sup-38_create_edit_delete_warehouse

svareyko commented on src/main/java/by/itech/lab/supplier/controller/WarehouseController.java:
```diff
@@ -6,17 +6,24 @@
-import static by.itech.lab.supplier.constant.ApiConstants.URL_WAREHOUSE;
+import javax.validation.Valid;
+
+import static by.itech.lab.supplier.constant.ApiConstants.*;
```
replace wildcard imports

svareyko commented on src/main/java/by/itech/lab/supplier/controller/WarehouseController.java:
```diff
@@ -30,4 +37,22 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public WarehouseDto save(@PathVariable(name = "customerId") Long customerId, @Valid @RequestBody WarehouseDto warehouseDto) {
```
same about variable name here

alexeyoverchenko replied:
I don't understand. what do you mean?

alexeyoverchenko committed:
- added entity address
- added AddressRepository
- correcter update function
- added http tests

alexeyoverchenko committed:
- added entity address
- added AddressRepository
- correcter update function
- added http tests

alexeyoverchenko committed:
- added save warehouse in users functional
- edited http tests

vareyko-itechart commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
-    public void delete(Long id) {
+    public WarehouseDto update(WarehouseDto warehouseDto) {
+        Warehouse entity = warehouseRepository.findById(warehouseDto.getId()).orElseThrow();
+        if (!warehouseDto.getIdentifier().equals(entity.getIdentifier())) {
```
![image](https://user-images.githubusercontent.com/44436165/98951102-e8c41f80-250a-11eb-9b56-2f97d9f872cf.png)

vareyko-itechart replied:
![Uploading image.png…]()

svareyko replied:
![image](https://user-images.githubusercontent.com/44436165/98951220-06918480-250b-11eb-8446-13c3e35649dd.png)

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -1,21 +1,40 @@
+
 
     @Override
     public Page<WarehouseDto> findAll(Pageable pageable) {
```
please, make parameter final

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
     @Override
-    public WarehouseDto save(WarehouseDto dto) {
-        return null;
+    public WarehouseDto save(Long customerId, WarehouseDto warehouseDto) {
```
please, make parameter final
Why is method called save if there is method update later. Can you unite two methods?

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
+        warehouseRepository.delete(id);
+    }
+
+    @Override
+    public WarehouseDto save(WarehouseDto dto) {
+        return null;
     }
```
If method does nothing, why do we need it?
You can put customer_id field into warehouseDto

alexeyoverchenko replied:
I should override this method, cuz it's in BaseService, when I use my own save method.

FortuneAwaker replied:
Soooo, maybe you should change your own method to make it override method in BaseService?

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
                 .orElseThrow(() -> new ResourceNotFoundException("Warehouse with id=" + warehouseId + " doesn't exist"));
```
please, make parameter final

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
+    }
+
+    @Override
+    public WarehouseDto save(WarehouseDto dto) {
```
please, make parameter final

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
-    public WarehouseDto save(WarehouseDto dto) {
-        return null;
+    public WarehouseDto save(Long customerId, WarehouseDto warehouseDto) {
+        if (warehouseDto.getIdentifier() == null) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_ACCEPTABLE, "\"Identifier\" can't be null");
+        }
```
We can put @Column(nullable = false) in entity class to avoid setting null field and getting exception message.
In another hand we usually use Objects.nonNull().

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/dto/mapper/WarehouseMapper.java:
```diff
@@ -21,8 +22,7 @@ public Warehouse map(WarehouseDto dto) {
                 .identifier(dto.getIdentifier())
```
please, make parameter final

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/dto/mapper/WarehouseMapper.java:
```diff
@@ -34,9 +34,15 @@ public WarehouseDto map(Warehouse entity) {
                 .identifier(entity.getIdentifier())
```
please, make parameter final

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
+                    HttpStatus.NOT_ACCEPTABLE, "\"Identifier\" can't be null");
+        }
+
+        Warehouse warehouse = warehouseMapper.map(warehouseDto);
+        Address savedAddress = addressCreate(warehouseDto.getAddressDto());
+        warehouse.setAddress(savedAddress);
```
How can you create warehouse with non-existing address? Address should be pre-created, shouldn't it?

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
+    @Transactional
     @Override
-    public void delete(Long id) {
+    public WarehouseDto update(WarehouseDto warehouseDto) {
```
please, make parameter final

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/dto/WarehouseDto.java:
```diff
@@ -26,9 +27,8 @@
-    private Address address;
-    private CustomerDto customerDto;
+    private AddressDto addressDto;
+    @OneToMany(mappedBy = "warehouse")
     private Set<UserDto> usersDto = new HashSet<>();
 }
```
Are you sure, that warehouse must know about users, that are connected to him?

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
+                    HttpStatus.NOT_ACCEPTABLE, "You can't change \"identifier\" field");
+        }
+
+        final Address existingAddress = addressRepository.findById(entity.getAddress().getId()).orElseThrow();
+        warehouseDto.getAddressDto().setId(existingAddress.getId());
+        addressMapper.map(warehouseDto.getAddressDto(), existingAddress);
+        addressRepository.save(existingAddress);
```
why you change addresses in warehouse?

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
-    public void delete(Long id) {
+    public WarehouseDto update(WarehouseDto warehouseDto) {
+        Warehouse entity = warehouseRepository.findById(warehouseDto.getId()).orElseThrow();
+        if (!warehouseDto.getIdentifier().equals(entity.getIdentifier())) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_ACCEPTABLE, "You can't change \"identifier\" field");
+        }
```
why can't we change identifier of warehouse? we only know that it should be unique.

FortuneAwaker commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
+        return warehouseMapper.map(savedWarehouse);
+    }
+
+    private Address addressCreate(AddressDto addressDto) {
```
please, make parameter final

alexeyoverchenko committed:
- fix after review

alexeyoverchenko committed:
- edited http file

alexeyoverchenko committed:
- removed wild import
- removed useful annotations

alexeyoverchenko committed:
Merge branch 'develop' into sup-38_create_edit_delete_warehouse

# Conflicts:
#	src/main/java/by/itech/lab/supplier/dto/mapper/WarehouseMapper.java

alexeyoverchenko committed:
- added getAllDispatchers method in User
- redone save and update functional
- optimized save and update

alexeyoverchenko committed:
Merge branch 'develop' into sup-38_create_edit_delete_warehouse

alexeyoverchenko committed:
- merged last develop

alexeyoverchenko committed:
- removed extra field

alexeyoverchenko committed:
- changed annotation

svareyko commented on src/main/java/by/itech/lab/supplier/controller/pre_tests.http:
```diff
@@ -77,53 +77,77 @@ Cookie: {{test}}
+Cookie: {{test}}
 
 ###
+#----------------------------------------------------------------------------------------------------------------------
```
?

alexeyoverchenko replied:
we use this to logically separate our classes. and its easier to navigate

svareyko replied:
look below, there is a comment inside of this separator

svareyko commented on src/main/java/by/itech/lab/supplier/repository/UserRepository.java:
```diff
@@ -31,4 +35,11 @@
     Page<User> findAllByActiveEquals(Pageable pageable, Boolean active);
 
+    @Modifying
+    @Query("update User set warehouse = :warehouse where id IN :usersId")
```
please dont use different case for operators, if you use uppercase - so use it everywhere, the same for lower case

alexeyoverchenko replied:
thanks

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -1,42 +1,79 @@
+        Customer customer = customerRepository.findById(warehouseDto.getCustomerId()).orElseThrow();
+        warehouse.setCustomer(customer);
+        Warehouse saved = warehouseRepository.save(warehouse);
+        userRepository.setWarehouseIntoUser(saved, warehouseDto.getUsersId());
```
dont access directly repositories that not related to your service

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -1,42 +1,79 @@
+import by.itech.lab.supplier.domain.Warehouse;
 import by.itech.lab.supplier.dto.WarehouseDto;
+import by.itech.lab.supplier.dto.mapper.AddressMapper;
+import by.itech.lab.supplier.dto.mapper.UserMapper;
```
remove unused import

alexeyoverchenko replied:
thanks

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -1,42 +1,79 @@
+    private final AddressRepository addressRepository;
+    private final UserRepository userRepository;
     private final WarehouseMapper warehouseMapper;
+    private final AddressMapper addressMapper;
```
if the address has its own mapper - maybe it should have its own service?

alexeyoverchenko replied:
we don't need service class for Address, cuz this information write and update through warehouse only

svareyko replied:
in any case it has its own repo, mapper, entity, dto, if you going to manage addresses in the scope of the warehouse - you dont need a repo and changes must be applied together with the warehouse entity

alexeyoverchenko committed:
- fixed

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java:
```diff
@@ -1,42 +1,74 @@
-    public void delete(Long id) {
+    private Warehouse create(final WarehouseDto warehouseDto) {
+        Warehouse warehouse = warehouseMapper.map(warehouseDto);
+        warehouse.getCustomer().setId(warehouseDto.getCustomerId());
```
maybe will be better to place this in the mapper?

alexeyoverchenko committed:
- added new map in CustomerMapper

alexeyoverchenko committed:
- cleaned code
- deleted AddressRepository
- "final" attribute in map

alexeyoverchenko committed:
- redacted http file
